datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategyFactory.java

```diff
@@ -0,0 +1,58 @@
+/**
+ *  Copyright 2021 LinkedIn Corporation. All rights reserved.
+ *  Licensed under the BSD 2-Clause License. See the LICENSE file in the project root for license information.
+ *  See the NOTICE file in the project root for additional information regarding copyright ownership.
+ */
+package com.linkedin.datastream.server.assignment;
+
+import java.util.Optional;
+import java.util.Properties;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linkedin.datastream.common.zk.ZkClient;
+import com.linkedin.datastream.server.DatastreamSourceClusterResolver;
+import com.linkedin.datastream.server.DummyDatastreamSourceClusterResolver;
+import com.linkedin.datastream.server.api.strategy.AssignmentStrategy;
+import com.linkedin.datastream.server.providers.NoOpPartitionThroughputProvider;
+import com.linkedin.datastream.server.providers.PartitionThroughputProvider;
+
+
+/**
+ * A factory for creating {@link LoadBasedPartitionAssignmentStrategy} instances
+ */
+public class LoadBasedPartitionAssignmentStrategyFactory extends StickyPartitionAssignmentStrategyFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(LoadBasedPartitionAssignmentStrategyFactory.class.getName());
+
+  @Override
+  public AssignmentStrategy createStrategy(Properties assignmentStrategyProperties) {
+    PartitionAssignmentStrategyConfig config = new PartitionAssignmentStrategyConfig(assignmentStrategyProperties);
+
+    boolean enableElasticTaskAssignment = config.isElasticTaskAssignmentEnabled();
+    // Create the zookeeper client
+    Optional<ZkClient> zkClient = Optional.empty();
+    try {
+      zkClient = constructZooKeeperClient(enableElasticTaskAssignment, config.getZkAddress(),
```

Comment:
this creation should happen only if `enableElastTaskAssignment` is set to true. Also, I think this can be simplified as all the parameters are present in config in the base class itself.
Base class can have something like this
`Optional<ZkClient> createZkClient() {
         if (config.isElasticTaskAssignmentEnabled()) {
             zkClient = construct ....
         }
     return zkClient
}`

Reply:
That's what happens @vmaheshw. Here's the method from the base class:

```
protected Optional<ZkClient> constructZooKeeperClient(boolean enableElasticTaskAssignment, String zkAddress,
      int zkSessionTimeout, int zkConnectionTimeout) {
    Optional<ZkClient> zkClient = Optional.empty();
    if (enableElasticTaskAssignment && StringUtils.isBlank(zkAddress)) {
      LOG.warn("ZkAddress is not present or empty");
      throw new IllegalStateException("ZkAddress is empty or not provided");
    }

    if (enableElasticTaskAssignment) {
      zkClient = Optional.of(new ZkClient(zkAddress, zkSessionTimeout, zkConnectionTimeout));
    }

    return zkClient;
  }
```

Reply:
Right, Then there is no point of retrieving the values in both the places. Your constructZkClient method should work without passing any field and using the existing config object itself.

You should not need to tweak enableElasticTaskAssignment in both base and derived class.